internal/auth/token.go

```diff
@@ -0,0 +1,190 @@
+		return "", fmt.Errorf("login request: %v", err)
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode < http.StatusOK || res.StatusCode >= http.StatusBadRequest {
+		return "", fmt.Errorf("login failed: %s", res.Status)
+	}
+
+	var tokenResponse struct {
+		Token string `json:"token"`
+	}
+	if err := json.NewDecoder(res.Body).Decode(&tokenResponse); err != nil {
+		return "", fmt.Errorf("decode token response: %v", err)
+	}
+
+	// Parse token expiry
+	claims, err := getClaims(tokenResponse.Token)
+	if err != nil {
+		return "", fmt.Errorf("parse token claims: %v", err)
+	}
+	if claims.Expiry == nil {
+		return "", fmt.Errorf("token does not contain expiry")
+	}
+
+	// Cache the token
+	p.token = tokenResponse.Token
+	p.tokenExpiry = claims.Expiry.Time()
+
+	return p.token, nil
+}
+
+func (p *LoginTokenProvider) Username() string {
+	return p.username
+}
+
+// AccessTokenProvider uses access tokens directly from the credential store
+type AccessTokenProvider struct {
+	configKey      string
+	cachedUsername string
+	configStore    *ConfigStore
+	mu             sync.Mutex
+}
+
+// NewAccessTokenProvider creates a token provider that uses access tokens from the credential store
+func NewAccessTokenProvider(configStore *ConfigStore, configKey string) *AccessTokenProvider {
+	return &AccessTokenProvider{
+		configKey:   configKey,
+		configStore: configStore,
+	}
+}
+
+func (p *AccessTokenProvider) EnsureToken(ctx context.Context) (string, error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	// Always get fresh access token from store (no caching for access tokens)
+	username, accessToken, err := p.configStore.GetCredentialStoreAccessTokens(p.configKey)
+	if err != nil {
+		return "", fmt.Errorf("get access token from store: %v", err)
+	}
+
+	// Cache username for display purposes
+	p.cachedUsername = username
+
+	return accessToken, nil
+}
+
+func (p *AccessTokenProvider) Username() string {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	return p.cachedUsername
+}
+
+// Helper methods for creating token providers
+
+// NewAccessTokenProviderFromStore creates an AccessTokenProvider from a ConfigStore
+// Returns error if no valid access token is available
+func NewAccessTokenProviderFromStore(configStore *ConfigStore, configKey string) (TokenProvider, error) {
+	// Test if we can get a valid access token
+	_, _, err := configStore.GetCredentialStoreAccessTokens(configKey)
+	if err != nil {
+		return nil, fmt.Errorf("no valid access token available: %v", err)
+	}
+
+	return NewAccessTokenProvider(configStore, configKey), nil
+}
+
+// NewLoginTokenProviderFromStore creates a LoginTokenProvider from pull credentials in the ConfigStore
+func NewLoginTokenProviderFromStore(configStore *ConfigStore, configKey, baseURL string) (TokenProvider, error) {
+	username, password, err := configStore.GetCredentialStorePullTokens(configKey)
+	if err != nil {
+		return nil, fmt.Errorf("no pull credentials available: %v", err)
+	}
+
+	if username == "" || password == "" {
```

Comment:
suggestion: Split out into separate checks to give better errors around which one is empty for whatever reason.